<a href="https://colab.research.google.com/github/BernardOforiBoateng/DSCI401/blob/main/DSCI401_HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Importing the pandas library
import pandas as pd

In [ ]:
# loading the package and tables
Batting = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Batting.csv")
People = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/People.csv")
Pitching = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Pitching.csv")

In [ ]:
Batting.head()

playerID  yearID  stint teamID lgID   G   AB   R   H  X2B  ...   RBI   SB  \
0  abercda01    1871      1    TRO  NaN   1    4   0   0    0  ...   0.0  0.0   
1   addybo01    1871      1    RC1  NaN  25  118  30  32    6  ...  13.0  8.0   
2  allisar01    1871      1    CL1  NaN  29  137  28  40    4  ...  19.0  3.0   
3  allisdo01    1871      1    WS3  NaN  27  133  28  44   10  ...  27.0  1.0   
4  ansonca01    1871      1    RC1  NaN  25  120  29  39   11  ...  16.0  6.0   

    CS  BB   SO  IBB  HBP  SH  SF  GIDP  
0  0.0   0  0.0  NaN  NaN NaN NaN   0.0  
1  1.0   4  0.0  NaN  NaN NaN NaN   0.0  
2  1.0   2  5.0  NaN  NaN NaN NaN   1.0  
3  1.0   0  2.0  NaN  NaN NaN NaN   0.0  
4  2.0   2  1.0  NaN  NaN NaN NaN   0.0  

[5 rows x 22 columns]

In [ ]:
Batting.columns

Index(['playerID', 'yearID', 'stint', 'teamID', 'lgID', 'G', 'AB', 'R', 'H',
       'X2B', 'X3B', 'HR', 'RBI', 'SB', 'CS', 'BB', 'SO', 'IBB', 'HBP', 'SH',
       'SF', 'GIDP'],
      dtype='object')

In [ ]:
People.head()

playerID  birthYear  birthMonth  birthDay birthCountry birthState  \
0  aardsda01     1981.0        12.0      27.0          USA         CO   
1  aaronha01     1934.0         2.0       5.0          USA         AL   
2  aaronto01     1939.0         8.0       5.0          USA         AL   
3   aasedo01     1954.0         9.0       8.0          USA         CA   
4   abadan01     1972.0         8.0      25.0          USA         FL   

    birthCity  deathYear  deathMonth  deathDay  ... weight height bats throws  \
0      Denver        NaN         NaN       NaN  ...  215.0   75.0    R      R   
1      Mobile     2021.0         1.0      22.0  ...  180.0   72.0    R      R   
2      Mobile     1984.0         8.0      16.0  ...  190.0   75.0    R      R   
3      Orange        NaN         NaN       NaN  ...  190.0   75.0    R      R   
4  Palm Beach        NaN         NaN       NaN  ...  184.0   73.0    L      L   

        debut   finalGame   retroID    bbrefID   deathDate   birthDate  
0  2004-04-06  2015-08-23  aardd001  aardsda01         NaN  1981-12-27  
1  1954-04-13  1976-10-03  aaroh101  aaronha01  2021-01-22  1934-02-05  
2  1962-04-10  1971-09-26  aarot101  aaronto01  1984-08-16  1939-08-05  
3  1977-07-26  1990-10-03  aased001   aasedo01         NaN  1954-09-08  
4  2001-09-10  2006-04-13  abada001   abadan01         NaN  1972-08-25  

[5 rows x 26 columns]

In [ ]:
People.columns

Index(['playerID', 'birthYear', 'birthMonth', 'birthDay', 'birthCountry',
       'birthState', 'birthCity', 'deathYear', 'deathMonth', 'deathDay',
       'deathCountry', 'deathState', 'deathCity', 'nameFirst', 'nameLast',
       'nameGiven', 'weight', 'height', 'bats', 'throws', 'debut', 'finalGame',
       'retroID', 'bbrefID', 'deathDate', 'birthDate'],
      dtype='object')

# (a)

## (10 points) Name every player in baseball history who has accumulated at least 300 home runs (HR) AND at least 300 stolen bases (SB). You can find the first and last name of the player in the People data frame. Join this to your result along with the total home runs and total bases stolen for each of these elite players.

In [ ]:
player_totals = Batting.loc[:,["playerID","HR","SB"]].groupby("playerID").sum()
players_over300 = player_totals[(player_totals['HR'] >= 300) & (player_totals['SB'] >= 300)]
elite_players = pd.merge(players_over300, People[["playerID", "nameFirst", "nameLast"]], on = "playerID", how = "inner")
elite_players = elite_players[['playerID', 'nameFirst', 'nameLast', 'HR', 'SB']]
elite_players

playerID nameFirst   nameLast   HR     SB
0  beltrca01    Carlos    Beltran  435  312.0
1  bondsba01     Barry      Bonds  762  514.0
2  bondsbo01     Bobby      Bonds  332  461.0
3  dawsoan01     Andre     Dawson  438  314.0
4  finlest01     Steve     Finley  304  320.0
5   mayswi01    Willie       Mays  660  338.0
6  rodrial01      Alex  Rodriguez  696  329.0
7  sandere02    Reggie    Sanders  305  304.0

# (b)

## (10 points) Similarly, name every pitcher in baseball history who has accumulated at least 300 wins (W) and at least 3,000 strikeouts (SO).

In [ ]:
Pitching.head()

playerID  yearID  stint teamID lgID   W   L   G  GS  CG  ...  IBB  WP  \
0  bechtge01    1871      1    PH1  NaN   1   2   3   3   2  ...  NaN   7   
1  brainas01    1871      1    WS3  NaN  12  15  30  30  30  ...  NaN   7   
2  fergubo01    1871      1    NY2  NaN   0   0   1   0   0  ...  NaN   2   
3  fishech01    1871      1    RC1  NaN   4  16  24  24  22  ...  NaN  20   
4  fleetfr01    1871      1    NY2  NaN   0   1   1   1   1  ...  NaN   0   

   HBP  BK     BFP  GF    R  SH  SF  GIDP  
0  NaN   0   146.0   0   42 NaN NaN   NaN  
1  NaN   0  1291.0   0  292 NaN NaN   NaN  
2  NaN   0    14.0   0    9 NaN NaN   NaN  
3  NaN   0  1080.0   1  257 NaN NaN   NaN  
4  NaN   0    57.0   0   21 NaN NaN   NaN  

[5 rows x 30 columns]

In [ ]:
pitcher_totals = Pitching.loc[:,["playerID","W","SO"]].groupby("playerID").sum()
players_over300_3000 = pitcher_totals[(pitcher_totals['W'] >= 300) & (pitcher_totals['SO'] >= 3000)]
elite_pitchers = pd.merge(players_over300_3000, People[["playerID", "nameFirst", "nameLast"]], on = "playerID", how = "inner")
elite_pitchers = elite_pitchers[['playerID', 'nameFirst', 'nameLast', 'W', 'SO']]
elite_pitchers

playerID nameFirst nameLast    W    SO
0  carltst01     Steve  Carlton  329  4136
1  clemero02     Roger  Clemens  354  4672
2  johnsra05     Randy  Johnson  303  4875
3  johnswa01    Walter  Johnson  417  3509
4  maddugr01      Greg   Maddux  355  3371
5  niekrph01      Phil   Niekro  318  3342
6  perryga01   Gaylord    Perry  314  3534
7   ryanno01     Nolan     Ryan  324  5714
8  seaveto01       Tom   Seaver  311  3640
9  suttodo01       Don   Sutton  324  3574

# (c)

## (10 points) Identify the name and year of every player who has hit at least 50 home runs in a single season. Which player had the lowest batting average in that season?


In [ ]:
# Identify the name and year of every player who has hit at least 50 home runs in a single season.
part_a = Batting.loc[:,["playerID", "yearID", "HR"]].groupby(["playerID", "yearID"],as_index = False).sum()
part_a = part_a[(part_a["HR"] >= 50)].sort_values("HR",ascending=False)
part_a.head()

playerID  yearID  HR
8450   bondsba01    2001  73
60024  mcgwima01    1998  70
86754   sosasa01    1998  66
60025  mcgwima01    1999  65
86757   sosasa01    2001  64

In [ ]:
# join players to People dataset
part_a = pd.merge(part_a, People[["playerID", "nameFirst", "nameLast"]], on="playerID", how = "inner")
part_a.head()

playerID  yearID  HR nameFirst nameLast
0  bondsba01    2001  73     Barry    Bonds
1  mcgwima01    1998  70      Mark  McGwire
2  mcgwima01    1999  65      Mark  McGwire
3  mcgwima01    1997  58      Mark  McGwire
4  mcgwima01    1996  52      Mark  McGwire

In [ ]:
# join players to Batting dataset and find the lowest batting average
part_b = pd.merge(part_a, Batting, on=("playerID","yearID"), how="inner")
part_b["batting_avg"]=part_b["H"]/part_b["AB"]
part_c = part_b[["playerID","yearID","nameFirst","nameLast","batting_avg"]].sort_values(by ="batting_avg", ascending = True)
part_c.head()

playerID  yearID nameFirst  nameLast  batting_avg
4   mcgwima01    1997      Mark   McGwire     0.252874
33  alonspe01    2019      Pete    Alonso     0.259631
32  bautijo02    2010      Jose  Bautista     0.260105
41  jonesan01    2005    Andruw     Jones     0.262799
10  marisro01    1961     Roger     Maris     0.269492

## Mark McGwire had the lowest batting average in that season with a batting average of 0.252874

In [ ]:
# loading the package and tables
flights = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/flights.csv")
planes = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/planes.csv")

In [ ]:
flights.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

In [ ]:
planes.head()

Unnamed: 0 tailnum    year                     type      manufacturer  \
0           1  N10156  2004.0  Fixed wing multi engine           EMBRAER   
1           2  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
2           3  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
3           4  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   
4           5  N10575  2002.0  Fixed wing multi engine           EMBRAER   

       model  engines  seats  speed     engine  
0  EMB-145XR        2     55    NaN  Turbo-fan  
1   A320-214        2    182    NaN  Turbo-fan  
2   A320-214        2    182    NaN  Turbo-fan  
3   A320-214        2    182    NaN  Turbo-fan  
4  EMB-145LR        2     55    NaN  Turbo-fan

# (a)

### (10 points) What is the oldest plane (specified by the tailnum variable) that flew from New York City airports in 2013?

In [ ]:
flights_0 = flights[["tailnum", "dep_time"]].dropna(subset=["dep_time"])
oldest_plane = pd.merge(flights_0, planes, on = "tailnum", how = "inner")
oldest_plane.sort_values(by = "year", ascending = True).head(5)


tailnum  dep_time  Unnamed: 0    year                     type  \
264146  N381AA    1505.0        1038  1956.0  Fixed wing multi engine   
264129  N381AA    1525.0        1038  1956.0  Fixed wing multi engine   
264130  N381AA    1737.0        1038  1956.0  Fixed wing multi engine   
264131  N381AA     817.0        1038  1956.0  Fixed wing multi engine   
264132  N381AA    1528.0        1038  1956.0  Fixed wing multi engine   

       manufacturer   model  engines  seats  speed         engine  
264146      DOUGLAS  DC-7BF        4    102  232.0  Reciprocating  
264129      DOUGLAS  DC-7BF        4    102  232.0  Reciprocating  
264130      DOUGLAS  DC-7BF        4    102  232.0  Reciprocating  
264131      DOUGLAS  DC-7BF        4    102  232.0  Reciprocating  
264132      DOUGLAS  DC-7BF        4    102  232.0  Reciprocating

## The oldest plane by tailnum is N381AA

# (b)

### (10 points) How many airplanes that flew from New York City are included in the planes table?

In [ ]:
flights_1 = set(oldest_plane.tailnum)
len(flights_1)

3316

In [ ]:
num_planes = planes['tailnum'].count()
num_planes

3322

### 3316 airplanes flew from NYC and is included in planes table.

# Question 3
### Generate the code to convert the following data frame to wide format.


In [ ]:
# Create the data frame
dat = pd.DataFrame({'grp': ['A', 'A', 'B', 'B'],
                   'sex': ['F', 'M', 'F', 'M'],
                   'meanL': [0.225, 0.47, 0.325, 0.547],
                   'sdL': [0.106, .325, .106, .308],
                   'meanR': [.34, .57, .4, .647],
                   'sdR': [0.0849, 0.325, 0.0707, 0.274]})
dat

grp sex  meanL    sdL  meanR     sdR
0   A   F  0.225  0.106  0.340  0.0849
1   A   M  0.470  0.325  0.570  0.3250
2   B   F  0.325  0.106  0.400  0.0707
3   B   M  0.547  0.308  0.647  0.2740

In [ ]:
dat_wide = pd.pivot(dat, index = "grp" , columns= "sex", values= ["meanL", "sdL", "meanR", "sdR"])
dat_wide

meanL           sdL        meanR            sdR       
sex      F      M      F      M     F      M       F      M
grp                                                        
A    0.225  0.470  0.106  0.325  0.34  0.570  0.0849  0.325
B    0.325  0.547  0.106  0.308  0.40  0.647  0.0707  0.274

# Question 4

###  Consider the pccc icd10 dataset.

In [ ]:
# loading the package and tables
pccc_icd10_dataset = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/pccc_icd10_dataset.csv")
pccc_icd10_dataset.head()

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7      dx8  \
0   1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G    O0973   
1   2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD  S72325C   
2   3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R      NaN   
3   4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A      NaN   
4   5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R  M8080XA   

       dx9  ...        g1        g2        g3        g4        g5        g6  \
0      NaN  ...  239196.0  672832.0  683784.0  757546.0       NaN  168052.0   
1  S52131B  ...  931331.0  404900.0  912213.0       NaN  964580.0  371556.0   
2  Y382X3D  ...  627455.0  638100.0  745829.0  843799.0  322975.0       NaN   
3      NaN  ...  809782.0  153243.0  413723.0  130995.0  211708.0  610135.0   
4  X731XXD  ...       NaN  636794.0       NaN  928572.0  930823.0  168586.0   

         g7        g8        g9       g10  
0  104625.0       NaN  850974.0       NaN  
1  778488.0  115827.0  440619.0  955264.0  
2       NaN  932106.0  289004.0  242699.0  
3       NaN  471383.0  191245.0  135116.0  
4  133292.0  699936.0  500743.0       NaN  

[5 rows x 31 columns]

### (a) Remove all the columns labeled with ”g” and a number

In [ ]:
# Remove all the columns labeled with "g" and a number
pccc_icd10_dataset_new = pccc_icd10_dataset.drop(columns=[col for col in pccc_icd10_dataset.columns if col.startswith("g")])
pccc_icd10_dataset_new.columns

Index(['id', 'dx1', 'dx2', 'dx3', 'dx4', 'dx5', 'dx6', 'dx7', 'dx8', 'dx9',
       'dx10', 'pc1', 'pc2', 'pc3', 'pc4', 'pc5', 'pc6', 'pc7', 'pc8', 'pc9',
       'pc10'],
      dtype='object')

### (b) Convert this to a long data set with three columns: id, type (pc or dx), and code.

In [ ]:
pccc_icd10_dataset_narrow = pd.melt(pccc_icd10_dataset_new, id_vars="id", value_vars=["dx1","dx2","dx3","dx4","dx5","dx6","dx7","dx8","dx9","dx10","pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10"],var_name="type",value_name="code")
pccc_icd10_dataset_narrow

id  type     code
0         1   dx1  S9410XS
1         2   dx1      NaN
2         3   dx1      NaN
3         4   dx1  S7226XK
4         5   dx1  S92246A
...     ...   ...      ...
19995   996  pc10  BR49ZZZ
19996   997  pc10  0T184KD
19997   998  pc10      NaN
19998   999  pc10  0N940ZZ
19999  1000  pc10  0NS5XZZ

[20000 rows x 3 columns]